rclcpp/include/rclcpp/time.hpp

```diff
@@ -33,38 +35,67 @@ class Time
   static Time
   now()
   {
-    rcutils_time_point_value_t rcutils_now = 0;
-    rcutils_ret_t ret = RCUTILS_RET_ERROR;
+    // TODO(karsten1987): This impl throws explicitely on RCL_ROS_TIME
+    // we have to do this, because rcl_time_source_init return RET_OK
+    // however sets system time under the hood
     if (ClockT == RCL_ROS_TIME) {
       throw std::runtime_error("RCL_ROS_TIME is currently not implemented.");
-      ret = false;
-    } else if (ClockT == RCL_SYSTEM_TIME) {
-      ret = rcutils_system_time_now(&rcutils_now);
-    } else if (ClockT == RCL_STEADY_TIME) {
-      ret = rcutils_steady_time_now(&rcutils_now);
     }
-    if (ret != RCUTILS_RET_OK) {
+
+    rcl_ret_t ret = RCL_RET_ERROR;
+
+    rcl_time_source_t time_source;
+    ret = rcl_time_source_init(ClockT, &time_source);
+
+    if (ret == RCL_RET_ERROR) {
       rclcpp::exceptions::throw_from_rcl_error(
-        ret, "Could not get current time: ");
+        ret, "could not initialize time source: ");
     }
 
-    return Time(std::move(rcutils_now));
+    if (ret == RCL_RET_INVALID_ARGUMENT) {
```

Comment:
What if ret is some other error like `RCL_RET_BAD_ALLOC` (just as an example)?

Wouldn't it better to move this check up and then do `if (ret != RCL_RET_OK) { rclcpp::exceptions::throw_from_rcl_error(...); }` after it?

See also: http://docs.ros2.org/beta2/api/rclcpp/namespacerclcpp_1_1exceptions.html#ad1d5563cfff9273f531a82c1dde13516